### PR TITLE
Yang

### PR DESCRIPTION
**Pull request by enzoliuyang94**
Edits related to issue #12

### CODE REVIEW

File: src/analysis/simulation.jl [hunk 1]
```diff
@@ -1,35 +1,35 @@
...
     wn = wntr.network.model.WaterNetworkModel(inpfilepath)
 
     # store pump names in a set
     pump_set = Set()
-    for (key,pump) in solution["1"]["pump"]
-        pump_name = data["pump"][key]["name"] # epanet name
+    for (key,pump) in wm_solution["solution"]["nw"]["1"]["pump"]
+        pump_name = wm_data["nw"]["1"]["pump"][key]["name"] # epanet name
         push!(pump_set, pump_name)
     end
```
Comment (tasseff) [inline src/analysis/simulation.jl:17]: You could replace this with something like `pump_set = [pump["name"] for (i, pump) in wm_data["nw"]["1"]["pump"]]`.

File: src/analysis/validation.jl [hunk 2]
```diff
@@ -2,102 +2,238 @@
...
+    head_loss_wntr = Array{Float64,1}(undef,duration)
+    head_loss_watermodels = Array{Float64,1}(undef,duration)
+
+    for t in 1:duration
+        flow_wntr[t] = wntr_simulation.link["flowrate"][link_name].values[t]
+        flow_watermodels[t] = wm_solution["solution"]["nw"][string(t)][link_type][link_id]["q"]
+        start_node = wntr_data.links._data[link_name].start_node_name
+        end_node = wntr_data.links._data[link_name].end_node_name
+        head_loss_wntr[t] = wntr_simulation.node["head"][start_node].values[t]-wntr_simulation.node["head"][end_node].values[t]
+        head_loss_watermodels[t] = -1*wm_solution["solution"]["nw"][string(t)][link_type][link_id]["dhn"]+wm_solution["solution"]["nw"][string(t)][link_type][link_id]["dhp"]
```
Comment (tasseff) [inline src/analysis/validation.jl:87]: `dhp` and `dhn` are only returned for relaxed problem formulations. This should be replaced with the true difference between heads at adjacent nodes using the `h` solution values.

File: src/analysis/simulation.jl [hunk 3]
```diff
@@ -44,41 +44,41 @@ function simulate(data::Dict{String,Any}, solution::Dict{String,Any},
...
+            node_to_index = string(wm_data["nw"]["1"][ltype][key]["node_to"])
+            node_fr_index = string(wm_data["nw"]["1"][ltype][key]["node_fr"])
 
-            node_to_name = data["node"][node_to_index]["source_id"][2]
-            node_fr_name = data["node"][node_fr_index]["source_id"][2]
+            node_to_name = wm_data["nw"]["1"]["node"][node_to_index]["source_id"][2]
+            node_fr_name = wm_data["nw"]["1"]["node"][node_fr_index]["source_id"][2]
 
             for tank_name in tank_set
                 if ((node_to_name == tank_name) & !(node_fr_name == tank_name)) | (!(node_to_name == tank_name) & (node_fr_name == tank_name))
```
Comment (tasseff) [inline src/analysis/simulation.jl:84]: Can we document what's happening here? Are you saying that if the link is not an artificial link, it should be added to `tank_link_dict`? Why is this necessary?

File: src/analysis/simulation.jl [hunk 4]
```diff
@@ -98,7 +98,7 @@ function simulate(data::Dict{String,Any}, solution::Dict{String,Any},
         diameter = wn.nodes._data[tank_name].diameter
 
         # enforce the initial levels of tanks 
-        wn.nodes._data[tank_name].init_level = solution["1"]["tank"][tank_index_dict[tank_name]]["V"]/(1/4*pi*diameter^2)
+        wn.nodes._data[tank_name].init_level = wm_solution["solution"]["nw"]["1"]["tank"][tank_index_dict[tank_name]]["V"]/(1/4*pi*diameter^2)
```
Comment (tasseff) [inline src/analysis/simulation.jl:101]: For safety, this should be computed as the head minus the elevation at the tank node instead of an operation on the volume. I'd advise using the pressure component of the solution, "p," which gives the same thing.

File: src/analysis/validation.jl [hunk 5]
```diff
@@ -2,102 +2,238 @@
...
-            if haskey(wntr_head, node_name)
-                node_head_wntr[node_name][tx] = wntr_head[node_name].values[tx]
-                node_head_wm[node_name][tx] = node["h"]
-            end
-        end
-    end
+# compare nodes
+function get_node_dataframe(wm_data,wm_solution,wntr_data,wntr_simulation, node_id)
+    duration = length(wm_solution["solution"]["nw"])            # number of time steps
+    time_step = wm_data["time_step"]/3600                       # length per time step (hour)
```
Comment (tasseff) [inline src/analysis/validation.jl:11]: There are other portions of the code (in `simulation.jl`) where you assume that the time step to be used in the EPANET simulation (particularly when adding controls) is one hour. You should use the same logic that you're using here to compute the actual times at which the controls should be set. For example, I've seen examples where the time step is 1800 seconds and not one hour.

File: src/analysis/validation.jl [hunk 6]
```diff
@@ -2,102 +2,238 @@
...
+    node_name = wm_data["nw"]["1"]["node"][node_id]["source_id"][2]
```
Comment (tasseff) [inline src/analysis/validation.jl:12]: I've noticed that in many locations, you are cross-referencing the EPANET data with the WaterModels data by comparing with either the field in [`"name"`] or `["source_id"][2]`. I'd advise using the latter everywhere throughout the code.

File: src/analysis/validation.jl [hunk 7]
```diff
@@ -2,102 +2,238 @@
...
-            node_name = data["node"][key]["source_id"][2]
-            if haskey(wntr_head, node_name)
-                node_head_wntr[node_name][tx] = wntr_head[node_name].values[tx]
-                node_head_wm[node_name][tx] = node["h"]
-            end
-        end
-    end
+# compare nodes
+function get_node_dataframe(wm_data,wm_solution,wntr_data,wntr_simulation, node_id)
+    duration = length(wm_solution["solution"]["nw"])            # number of time steps
```
Comment (tasseff) [inline src/analysis/validation.jl:10]: This should be called `num_steps` or something similar instead of `duration`. Duration should refer to something measured in seconds.

File: src/analysis/validation.jl [hunk 8]
```diff
@@ -2,102 +2,238 @@
...
-            # else
-            #     link_name = data[ltype][key]["name"]
-            # end
+    return node_df
+end
 
-            link_name = data[ltype][key]["name"]
+# compare tanks
+function get_tank_dataframe(wm_data,wm_solution,wntr_data,wntr_simulation, tank_id)
+    duration = length(wm_solution["solution"]["nw"])            # number of time steps
```
Comment (tasseff) [inline src/analysis/validation.jl:37]: Same comment as above regarding `duration`.

File: src/analysis/validation.jl [hunk 9]
```diff
@@ -2,102 +2,238 @@
...
-            link_flow_wm[link_name] = Array{Float64,1}(undef,length(solution))
+    for t in 1:duration
+        volume_wntr[t] = wntr_simulation.node["pressure"][tank_name].values[t]*(1/4*pi*diameter^2)
+        volume_watermodels[t] = wm_solution["solution"]["nw"][string(t)]["tank"][tank_id]["V"]
 
-            link_dh_wntr[link_name] = Array{Float64,1}(undef,length(solution))
-            link_dh_wm[link_name] = Array{Float64,1}(undef,length(solution))
-        end
+        level_wntr[t] = wntr_simulation.node["pressure"][tank_name].values[t]
+        level_watermodels[t] = wm_solution["solution"]["nw"][string(t)]["tank"][tank_id]["V"]/(1/4*pi*diameter^2)
```
Comment (tasseff) [inline src/analysis/validation.jl:52]: Again, I'd use `"p"` here at the corresponding tank node to compute the WaterModels level.

File: src/analysis/validation.jl [hunk 10]
```diff
@@ -2,102 +2,238 @@
...
     end
 
+    tank_df = DataFrame(time = 1:time_step:time_step*duration, volume_wntr = vec(volume_wntr), volume_watermodels = vec(volume_watermodels),
```
Comment (tasseff) [inline src/analysis/validation.jl:55]: An aside, but why does everything need to be cast to type `vec`, here?

File: src/analysis/validation.jl [hunk 11]
```diff
@@ -2,102 +2,238 @@
...
+function get_link_dataframe(wm_data,wm_solution,wntr_data,wntr_simulation, link_id)
+    if link_id in keys(wm_data["nw"]["1"]["pipe"])
+        link_type = "pipe"
+    elseif link_id in keys(wm_data["nw"]["1"]["short_pipe"])
+        link_type = "short_pipe"
+    elseif link_id in keys(wm_data["nw"]["1"]["valve"])
+        link_type = "valve"
+    else
+        link_type = "pump"
+    end
```
Comment (tasseff) [inline src/analysis/validation.jl:71]: Add another condition for the "regulator" type.

File: src/analysis/validation.jl [hunk 12]
```diff
@@ -2,102 +2,238 @@
...
+    head_loss_wntr = Array{Float64,1}(undef,duration)
+    head_loss_watermodels = Array{Float64,1}(undef,duration)
+
+    for t in 1:duration
+        flow_wntr[t] = wntr_simulation.link["flowrate"][pipe_name].values[t]
+        flow_watermodels[t] = wm_solution["solution"]["nw"][string(t)]["pipe"][pipe_id]["q"]
+        start_node = wntr_data.links._data[pipe_name].start_node_name
+        end_node = wntr_data.links._data[pipe_name].end_node_name
+        head_loss_wntr[t] = wntr_simulation.node["head"][start_node].values[t]-wntr_simulation.node["head"][end_node].values[t]
+        head_loss_watermodels[t] = -1*wm_solution["solution"]["nw"][string(t)]["pipe"][pipe_id]["dhn"]+wm_solution["solution"]["nw"][string(t)]["pipe"][pipe_id]["dhp"]
```
Comment (tasseff) [inline src/analysis/validation.jl:112]: Same comment as above when computing head differences from the WaterModels solution.

File: src/analysis/validation.jl [hunk 13]
```diff
@@ -2,102 +2,238 @@
...
+    head_loss_wntr = Array{Float64,1}(undef,duration)
+    head_loss_watermodels = Array{Float64,1}(undef,duration)
+
+    for t in 1:duration
+        flow_wntr[t] = wntr_simulation.link["flowrate"][short_pipe_name].values[t]
+        flow_watermodels[t] = wm_solution["solution"]["nw"][string(t)]["short_pipe"][short_pipe_id]["q"]
+        start_node = wntr_data.links._data[short_pipe_name].start_node_name
+        end_node = wntr_data.links._data[short_pipe_name].end_node_name
+        head_loss_wntr[t] = wntr_simulation.node["head"][start_node].values[t]-wntr_simulation.node["head"][end_node].values[t]
+        head_loss_watermodels[t] = -1*wm_solution["solution"]["nw"][string(t)]["short_pipe"][short_pipe_id]["dhn"]+wm_solution["solution"]["nw"][string(t)]["short_pipe"][short_pipe_id]["dhp"]
```
Comment (tasseff) [inline src/analysis/validation.jl:137]: Same comment as above when computing head differences for the WaterModels solution.

File: src/analysis/validation.jl [hunk 14]
```diff
@@ -2,102 +2,238 @@
...
+    head_loss_wntr = Array{Float64,1}(undef,duration)
+    head_loss_watermodels = Array{Float64,1}(undef,duration)
+
+    for t in 1:duration
+        flow_wntr[t] = wntr_simulation.link["flowrate"][valve_name].values[t]
+        flow_watermodels[t] = wm_solution["solution"]["nw"][string(t)]["valve"][valve_id]["q"]
+        start_node = wntr_data.links._data[valve_name].start_node_name
+        end_node = wntr_data.links._data[valve_name].end_node_name
+        head_loss_wntr[t] = wntr_simulation.node["head"][start_node].values[t]-wntr_simulation.node["head"][end_node].values[t]
+        head_loss_watermodels[t] = -1*wm_solution["solution"]["nw"][string(t)]["valve"][valve_id]["dhn"]+wm_solution["solution"]["nw"][string(t)]["valve"][valve_id]["dhp"]
```
Comment (tasseff) [inline src/analysis/validation.jl:162]: Same comment as above when computing head differences for the WaterModels solution.

File: src/analysis/validation.jl [hunk 15]
```diff
@@ -2,102 +2,238 @@
...
+    end
+
+    for t in 1:duration
+        status[t] = wm_solution["solution"]["nw"][string(t)]["pump"][pump_id]["status"]
+        flow_wntr[t] = wntr_simulation.link["flowrate"][pump_name].values[t]
+        flow_watermodels[t] = wm_solution["solution"]["nw"][string(t)]["pump"][pump_id]["q"]
+        start_node = wntr_data.links._data[pump_name].start_node_name
+        end_node = wntr_data.links._data[pump_name].end_node_name
+        # the head gain here measures the head difference, not necessarily the lift of an operating pump
+        head_gain_wntr[t] = wntr_simulation.node["head"][end_node].values[t]-wntr_simulation.node["head"][start_node].values[t]
```
Comment (tasseff) [inline src/analysis/validation.jl:218]: Ensure that if the pump is off, the head gain is set to zero, here.

File: src/analysis/validation.jl [hunk 16]
```diff
@@ -2,102 +2,238 @@
...
+        head_gain_watermodels[t] = wm_solution["solution"]["nw"][string(t)]["pump"][pump_id]["dhn"]-wm_solution["solution"]["nw"][string(t)]["pump"][pump_id]["dhp"]
```
Comment (tasseff) [inline src/analysis/validation.jl:219]: Same comment as above when computing head differences for the WaterModels solution.

File: src/analysis/validation.jl [hunk 17]
```diff
@@ -2,102 +2,238 @@
...
+                    break
                 end
             end
+        elseif wntr_data.options.energy.global_efficiency != nothing
+            eff = wntr_data.options.energy.global_efficiency/100 # e.g. convert 60 to 60% (0.6)
+        else
+            println("No pump efficiency provided, 75% is used")
+            eff = 0.75
+        end
+        power = 9.81*dh*q/eff # kilowatt (kW)
```
Comment (tasseff) [inline src/analysis/validation.jl:207]: Report back the power in Watts instead of kilowatts.

File: src/analysis/visualization.jl [hunk 18]
```diff
@@ -5,21 +5,22 @@ Compare the fluctuations in tank levels calculated from WaterModels and WNTR
...
-    p = plot!(p,duration,node_head_wm[tank_name].-wn.nodes._data[tank_name].elevation,label="WaterModel")
-
-	savefig(p,outfilepath*tank_name*".png")
-
-    println("----- Tank level plot saved -----")
-  
-    return true
+    for t in 1:duration
+        level_wntr[t] = wntr_simulation.node["pressure"][tank_name].values[t]
+        level_watermodels[t] = wm_solution["solution"]["nw"][string(t)]["tank"][tank_id]["V"]/(1/4*pi*diameter^2)
```
Comment (tasseff) [inline src/analysis/visualization.jl:17]: Use the pressure at the node to which the tank is attached.

**Review by tasseff — COMMENTED** [review]
Thanks Yang! I've completed a short review. In general, I think the validation functions look really nice. In the future, either through more documentation or through replicating the WaterModels network in WNTR by constructing it from the ground up, we could greatly improve the understandability of the simulation setup code.

File: src/analysis/simulation.jl (hunk 3 above)
Comment (enzoliuyang94) [reply src/analysis/simulation.jl:84]: @tasseff  Here, artificial links refer to those zero-length pipes with shutoff valves added to tanks, and links stored in tank_link_dict are those links that are connected to the tank in the absence of artificial links. I think this part of the codes can be simplified - it looks messy now as I modified it two weeks ago when the artificial node and tank node were reversed in WaterModels. I will try to simplify the codes here and write down some notes next to it.

**Comment by enzoliuyang94** [comment]
@tasseff  Will implement the changes you suggest. Thanks.

**Comment by tasseff** [comment]
Found an edge case while running this in an iterative algorithm: `ValueError('4.624143259902508e-09 is not a valid LinkStatus')`. I think this can be resolved by, in your simulation code, using `shutoff_valve_status = round(shutoff_valve_dict["status"])` and `pump_status = round(pump_dict["status"])`.

**Comment by enzoliuyang94** [comment]
> Found an edge case while running this in an iterative algorithm: `ValueError('4.624143259902508e-09 is not a valid LinkStatus')`. I think this can be resolved by, in your simulation code, using `shutoff_valve_status = round(shutoff_valve_dict["status"])` and `pump_status = round(pump_dict["status"])`.

@tasseff Will address this as well as the other listed issues. Thanks for pointing this out.

**Comment by tasseff** [comment]
@enzoliuyang94: Another error I just discovered running an algorithm on the "test/data/epanet/multinetwork/owf-hw-lps.inp" example in WaterModels: `ValueError('The name provided for the control is already used. Please either remove the control with that name first or use a different name for this control.')`. Not sure why this is.

**Comment by enzoliuyang94** [comment]
> @enzoliuyang94: Another error I just discovered running an algorithm on the "test/data/epanet/multinetwork/owf-hw-lps.inp" example in WaterModels: `ValueError('The name provided for the control is already used. Please either remove the control with that name first or use a different name for this control.')`. Not sure why this is.

@tasseff After looking at the naming convention I used on the controls, one reason I can think of that a shutoff valve happen to have the same name as a pump - under the current codes, this would result in same control names. I will adjust the naming convention for the controls to avoid this issue.

**Comment by enzoliuyang94** [comment]
> > @enzoliuyang94: Another error I just discovered running an algorithm on the "test/data/epanet/multinetwork/owf-hw-lps.inp" example in WaterModels: `ValueError('The name provided for the control is already used. Please either remove the control with that name first or use a different name for this control.')`. Not sure why this is.
> 
> @tasseff After looking at the naming convention I used on the controls, one reason I can think of that a shutoff valve happen to have the same name as a pump - under the current codes, this would result in same control names. I will adjust the naming convention for the controls to avoid this issue.

@tasseff  I just modified the codes to address the error messages you encounter, including the fragmentation error. Do you think it's possible that Julia is doing some parallel processing itself? I am currently generating random names for the temporary inp files that is generated when doing the hydraulic simulation, if there is indeed some underlying parallel processing, we may need to use separate random seeds to generate the random numbers.

**Comment by jjstickel** [comment]
I got setup to run your test script and make my review, and also found that DataFrames needs to be added to Project.toml. I'll get back to it on Monday. If you've been making edits in response to Byron's comments, please make a commit to the branch.

File: src/analysis/validation.jl (hunk 10 above)
Comment (enzoliuyang94) [reply src/analysis/validation.jl:55]: If I don't add 'vec', there would be an error building the dataframe

File: src/WaterModelsAnalytics.jl [hunk 19]
```diff
@@ -16,8 +16,8 @@ module WaterModelsAnalytics
     using Printf
 
     import Plots
+    import DataFrames
```
[the remark this answers is not in the document]
Comment (jjstickel) [reply src/WaterModelsAnalytics.jl:19]: Yang did this.

File: src/analysis/simulation.jl (hunk 1 above)
Comment (jjstickel) [reply src/analysis/simulation.jl]: Looks like Yang made this change

File: src/analysis/simulation.jl (hunk 4 above)
Comment (jjstickel) [reply src/analysis/simulation.jl]: Not sure why computing something from the volume shouldn't work consistently, but this has been changed now. Although I am making an update to use the correct node.

Comment (enzoliuyang94) [reply src/analysis/simulation.jl]: I noticed that using the tank node's pressure will give the wrong (impossible) tank level sometimes - for example, the pressure of a tank node solved by WaterModels could be negative. However, the pressure of the artificial node added next to the tank seems to be always consistent with the real tank level that is calculated from the tank volume. Therefore, using 'pressure' to compute the tank level actually now involves more codes as I need to first identify the 'artificial node' attached to the tank. 

I do think using volume is probably more straightforward.

Comment (jjstickel) [reply src/analysis/simulation.jl]: I think this was when the artificial node and tank were reversed. I am about to make a commit that uses the pressure for the tank-node. No negative pressure and levels.

**Comment by jjstickel** [comment]
I'm partway through. The code runs, I verified that the modified WNTR network has the right structure, and I fixed an issue with calculating tank levels. I hope to finish tomorrow.

File: src/analysis/validation.jl (hunk 2 above)
Comment (jjstickel) [reply src/analysis/validation.jl]: Looks like Yang resolved this.

File: src/analysis/simulation.jl (hunk 3 above)
Comment (jjstickel) [reply src/analysis/simulation.jl:73]: @tasseff , are Yang's revisions sufficient to satisfy you?

File: src/analysis/validation.jl (hunk 5 above)
Comment (jjstickel) [reply src/analysis/validation.jl:11]: Looks like Yang has done this.

File: src/analysis/validation.jl (hunk 7 above)
Comment (jjstickel) [reply src/analysis/validation.jl]: Looks like this has been fixed.

File: src/analysis/validation.jl (hunk 6 above)
Comment (jjstickel) [reply src/analysis/validation.jl:12]: I see that Yang changed some of these; I'll presume he got them all.

File: src/analysis/validation.jl (hunk 10 above)
Comment (jjstickel) [reply src/analysis/validation.jl]: @enzoliuyang94 , I am with @tasseff . If you declare these variables as `Array{Float64,1}` then they will already be vectors and you do not need `vec(var)`; it is redundant. I tested the removal of `vec()` in one case, and there was no error. Maybe there was one variable that caused the error?

File: src/analysis/validation.jl (hunk 15 above)
Comment (jjstickel) [reply src/analysis/validation.jl]: @enzoliuyang94, you multiplied both the wm and wntr solution head gains by the wm status. I think you should use the wntr status for the wntr head gain. This is one way to check that the WNTR network controls were defined properly. You can get it from `wntr_simulation.link["status"]["pmp1"].values[t]`.

Comment (enzoliuyang94) [reply src/analysis/validation.jl]: @jjstickel @tasseff I did what Jonathan suggested, and also create 'status_wntr' and 'status_watermodels' as two separate columns in the dataframe output.

**Comment by enzoliuyang94** [comment]
> @enzoliuyang94: Another error I just discovered running an algorithm on the "test/data/epanet/multinetwork/owf-hw-lps.inp" example in WaterModels: `ValueError('The name provided for the control is already used. Please either remove the control with that name first or use a different name for this control.')`. Not sure why this is.

@tasseff I think this is caused by the same 'naming issue' that I mentioned in the Email I just sent to you and Jonathan.

File: src/analysis/simulation.jl (hunk 4 above)
Comment (tasseff) [reply src/analysis/simulation.jl]: It is fine to use volume for now since we assume only cylindrical tanks, but this could change in the future if we decide to model non-cylindrical tanks.

File: src/analysis/simulation.jl (hunk 3 above)
Comment (tasseff) [reply src/analysis/simulation.jl:71]: The top comment that was added helps, thanks.

File: src/analysis/validation.jl (hunk 15 above)
Comment (jjstickel) [reply src/analysis/validation.jl]: LGTM now. Thanks.

**Comment by jjstickel** [comment]
@enzoliuyang94 , Thanks for the effort on this one. I think it is ready to merge -- go ahead if you don't have any loose ends.
